Rectify timeline in Touchbar freezes, #4058

low-batt opened the pull request:
This commit will:
- Add a new method makeTouchBar to PlayerCore
- Change MainWindowController and MiniPlayerWindowController extensions in TouchBarSupport to call the new method
- Add a hasTouchBar property to TouchBarSupport
- Change MainWindowController.hideUI to not stop the timer that synchronizes the UI if the Mac has a touch bar

This corrects a regression added by a change to hideUI to stop the timer when the OSC is hidden in order to save energy. That change failed to take into account that the timer also synchronizes controls in the touch bar on MacBooks that have one.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #4058.

---

**Description:**

low-batt committed:
Rectify timeline in Touchbar freezes, #4058

This commit will:
- Add a new method makeTouchBar to PlayerCore
- Change MainWindowController and MiniPlayerWindowController extensions
  in TouchBarSupport to call the new method
- Add a hasTouchBar property to TouchBarSupport
- Change MainWindowController.hideUI to not stop the timer that
  synchronizes the UI if the Mac has a touch bar

This corrects a regression added by a change to hideUI to stop the timer
when the OSC is hidden in order to save energy. That change failed to
take into account that the timer also synchronizes controls in the
touch bar on MacBooks that have one.

low-batt commented:
The regression was due to the change that stops the timer that synchronizes the UI when the OSC is hidden in order to save energy. That change failed to take into account that the timer also synchronizes controls in the touch bar on MacBooks that have one.

In trying to correct the regression and yet not run the timer when it is not needed I looked for APIs that would answer:
- Does the Mac have a touch bar?
- Is the touch bar configured to show app controls?
- Is the touch bar awake?
- Is the MacBook in closed clamshell mode?

Unfortunately Apple thinks there is no need for an application to know if the Mac has a touch bar. From the documentation for [NSTouchBar](https://developer.apple.com/documentation/appkit/nstouchbar):
> There’s no need, and no API, for your app to know whether or not there’s a Touch Bar available. Whether your app is running on a machine that supports the Touch Bar or not, your app’s onscreen user interface (UI) appears and behaves the same way.

I found [isVisible](https://developer.apple.com/documentation/appkit/nstouchbaritem/2544669-isvisible) was `true` even when the touch bar was not displaying the application's UI.

So to determine if the Mac has a touch bar the code is checking whether AppKit has called [makeTouchBar](https://developer.apple.com/documentation/appkit/nsresponder/2544690-maketouchbar). This allows the timer to be stopped on Macs that do not have touch bars. This also avoided running the timer when a MacBook with a touch bar is running in closed clamshell mode.

Reviewers, is there a better way to determine if a touch bar is present?

low-batt committed:
Rectify progress bar freezes playing any audio file, #4051

This commit will:
- Change MainWindowController.hideUI to not stop the timer that
  synchronizes the UI if displaying the mini player
- Change PlayerCore.switchToMiniPlayer to ensure the timer is running
  if media is playing

low-batt commented:
I've added an additional commit that addresses issue #4051 as it involves the same code.

saagarjha commented on iina/PlayerCore.swift:
```diff
@@ -93,6 +93,12 @@ class PlayerCore: NSObject {
+  ///
+  /// - Note: This is set based on whether `AppKit` has called `MakeTouchBar`, therefore it can, for example, be `false` for
+  ///         a MacBook that has a touch bar if the touch bar is asleep because the Mac is in closed clamshell mode.
+  var hasTouchBar = false
```
```suggestion
  var needsTouchBar = false
```

Based on the comment above.

low-batt replied:
Latest commit renames the new property to needsTouchBar as recommended.

low-batt committed:
Rectify timeline in Touchbar freezes, #4058

This commit will rename the new hasTouchBar property to needsTouchBar
as recommended by review comment.
